229: Adding account profile and email names to nav drawer

eliserichards opened the pull request:
Fixes #229 
Relates to #135 

## Testing and Review Notes
before
<img width="350" alt="image" src="https://user-images.githubusercontent.com/43795363/49315679-67049400-f4a3-11e8-89e2-bee1cd087c79.png">

after:
<img width="350" alt="image" src="https://user-images.githubusercontent.com/43795363/49843303-b47ade00-fd8c-11e8-8b90-f38d57bb1960.png">
<img width="350" alt="image" src="https://user-images.githubusercontent.com/43795363/49843308-b6dd3800-fd8c-11e8-84a6-b8800c3c31cd.png">

## To Do

- [x] add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
- [x] request the "UX" team perform a design review (if/when applicable)
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)
- [x] check on the [accessibility](https://mozilla-lockbox.github.io/lockbox-android/accessibility/) of any added UI

jhugman commented on app/src/main/java/mozilla/lockbox/presenter/ItemListPresenter.kt:
```diff
@@ -37,7 +37,9 @@ interface ItemListView {
-    fun setDisplayName(text: String)
+    fun setDisplayEmailName(text: String)
+    fun setAccountName(text: String)
+    fun setAvatarFromURL(url: String)
```
I think I'd prefer `var` properties here.  Setters like this are more reminiscent of idiomatic Java than Kotlin.

```java
public void setProperty(Property property) {}
```

:-P

jhugman commented on app/src/main/java/mozilla/lockbox/presenter/ItemListPresenter.kt:
```diff
@@ -106,7 +108,25 @@ class ItemListPresenter(
+            }
+            .filterNotNull()
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe(view::setAvatarFromURL)
```
This looks like three separate subscriptions to the same thing. Could you combine them?

e.g.

```kotlin
accountStore.profile
  .map { it.value?.asOptional() }
  .filterNotNull()
  .subscribe {
    view.avatarURL = it.avatar
    view.displayEmailName = it.email
    view.accountName = …
  }
```

Alternatively, you can use create a view model (not the Android architecture object) to pass all the objects up together.

jhugman commented on app/src/main/java/mozilla/lockbox/view/ItemListFragment.kt:
```diff
@@ -129,10 +131,25 @@ class ItemListFragment : CommonFragment(), ItemListView {
     }
 
+    override fun setAvatarFromURL(url: String) {
+        Picasso.get()
```
Ooh, fancy! :)

eliserichards replied:
thanks, I stole it from Sasha's impl ;)

jhugman commented on app/src/main/java/mozilla/lockbox/view/ItemListFragment.kt:
```diff
@@ -129,10 +131,25 @@ class ItemListFragment : CommonFragment(), ItemListView {
+            .load(url)
+            .placeholder(R.drawable.ic_avatar_placeholder)
+            .resize(80, 80)
+            .transform(CropCircleTransformation())
```
1. could we put the `80` someplace more accessible to designers, e.g. `res/attrs.xml`
2. `CropCircleTransformation` do we need a new one of these each time, or should we try and use only one. This is a minor point.

eliserichards replied:
1. Done!
2. Not sure, I will take a look at this

eliserichards commented:
@jhugman: I addressed your comments. I created a view model `AccountViewModel` that has the three attributes, then set them all in the same `accountStore.profile.map{...}.subscribe{...}`. Let me know what you think! 💃

jhugman commented on app/src/main/java/mozilla/lockbox/store/AccountStore.kt:
```diff
@@ -118,18 +119,20 @@ open class AccountStore(
     }
 
     private fun generateLoginURL() {
+        println("ELISETEST: generateLoginUrl hit")
```
Nit: cute logging :)

jhugman commented on app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt:
```diff
@@ -43,21 +45,37 @@ class FxALoginFragment : BackableFragment(), FxALoginView {
         if (!isDebug()) {
             skipFxA.visibility = View.GONE
         }
+
```
Nit: whitespace

jhugman commented on app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt:
```diff
@@ -43,21 +45,37 @@ class FxALoginFragment : BackableFragment(), FxALoginView {
+    }
+
     override fun loadURL(url: String) {
+        println("ELISETEST: loadUrl hit")
```
Nit: Happy to have logging, though you should probably use `log.debug()`. Best remove this sort of thing before committing :)

eliserichards replied:
`log.debug()`: noted! 👍 👍 👍

jhugman commented on app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt:
```diff
@@ -43,21 +45,37 @@ class FxALoginFragment : BackableFragment(), FxALoginView {
+            override fun onReceivedHttpError(
+                view: WebView?,
+                request: WebResourceRequest?,
+                errorResponse: WebResourceResponse?
```
OMFG Are we going to get an error page? Yipppeee!

jhugman commented on app/src/main/res/values/attrs.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <attr name="avatar_image_size">80dp</attr>
+</resources>
```
I got this wrong in the last review; it should be a `dimen`, not an `attr`. My apologies.

jhugman commented on app/src/main/java/mozilla/lockbox/view/ItemListFragment.kt:
```diff
@@ -171,6 +173,17 @@ class ItemListFragment : CommonFragment(), ItemListView {
+        Picasso.get()
+            .load(profile.avatarFromURL)
+            .placeholder(R.drawable.ic_avatar_placeholder)
+            .resize(R.attr.avatar_image_size, R.attr.avatar_image_size)
```
Oooh. This might not work the way you expect it. `R.attr.avatar_image_size` is a generated id, rather than the value itself.

```kotlin
Picasso.with(context)
    .load(url)
    .placeholder(R.drawable.whatever)
    .resizeDimen(R.dimen.avatar_image_size, R.dimen.avatar_image_size)
```

(I realize now I suggested you might move this into a `attrs.xml` file, and now I remember it should be in a `dimens.xml` file. Apologies.)

jhugman replied:
You may also want to use a `fit` method, but I haven't read those docs too closely.

eliserichards replied:
done

jhugman commented on app/src/main/java/mozilla/lockbox/view/ItemListFragment.kt:
```diff
@@ -171,6 +173,17 @@ class ItemListFragment : CommonFragment(), ItemListView {
         adapter.updateItems(itemList)
     }
 
+    override fun updateAccountProfile(profile: AccountViewModel) {
```
So much nicer.

jhugman commented on app/src/main/java/mozilla/lockbox/model/AccountViewModel.kt:
```diff
@@ -0,0 +1,13 @@
+    val accountName: String,
+    val displayEmailName: String,
+    val avatarFromURL: String
+)
```
Your source model is full of optionals, so I'd make these optional too.

Then leave the choice of how to display (including the placeholder) up to the view/fragment layer.

eliserichards replied:
I'm a little confused about how to make these optional. Also struggling with the styling suggestion (`map`/`filterNotNull`/`subscribe`), I think I'm using `apply` wrong but I can't quite figure it out. Maybe we can talk on Friday or I can pull Sasha in for a little pair programming :)

eliserichards replied:
I spoke to Sasha and she was able to clarify :D

jhugman commented on app/src/main/java/mozilla/lockbox/presenter/ItemListPresenter.kt:
```diff
@@ -98,17 +102,23 @@ class ItemListPresenter(
+                    avatarFromURL = profile.value.avatar!!
+                )
+                view.updateAccountProfile(account)
+            }
```
This is so much nicer.

From a style pov, I'm leaning towards a `map`/`filterNotNull`/`subscribe` pipeline.

e.g. 

```kotlin
.map { profile -> 
     profile?.value?.apply {
         AccountModel(displayName ?: email, email, avatar)
     }.asOptional()
}
.filterNotNull()
.subscribe(view::updateAccountProfile)
```

eliserichards replied:
done

sashei commented on app/src/main/java/mozilla/lockbox/presenter/ItemListPresenter.kt:
```diff
@@ -53,36 +53,40 @@ class ItemListPresenter(
-                .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(view::updateItemListSort)
-                .addTo(compositeDisposable)
+            .distinctUntilChanged()
```
👾

sashei commented on app/src/test/java/mozilla/lockbox/presenter/ItemListPresenterTest.kt:
```diff
@@ -102,8 +106,8 @@ open class ItemListPresenterTest {
-        override fun setDisplayName(text: String) {
-            setDisplayNameArgument = text
+        override fun updateAccountProfile(profile: AccountViewModel) {
+            setDisplayNameArgument = profile.displayEmailName
```
it might be good to test the `accountName` property in both `ServerProfile.displayName` cases (null vs. non-null).

eliserichards replied:
agreed. on my todo list for tomorrow

sashei commented on app/src/main/java/mozilla/lockbox/presenter/ItemListPresenter.kt:
```diff
@@ -98,17 +102,21 @@ class ItemListPresenter(
+            }.addTo(compositeDisposable)
 
         accountStore.profile
+            .filterNotNull()
```
😍

nickbrandt commented:
> @nickbrandt : What would you prefer to happen if a user has no profile pic/display name?

In the case the user does not have a display name or avatar, we would use the Lockbox icon that we had previously along with "Firefox Lockbox" where the display name is. (with their email below it)

eliserichards commented:
@nickbrandt I added some screenshots up top -- let me know what you think (I'm not liking the white outline around the lockbox logo...)

jhugman commented on app/src/main/java/mozilla/lockbox/view/FxALoginFragment.kt:
```diff
@@ -40,9 +43,9 @@ class FxALoginFragment : BackableFragment(), FxALoginView {
-        if (!isDebug()) {
+        if (!isDebug())
             skipFxA.visibility = View.GONE
-        }
```
Nit: keep braces, even for one line blocks.

eliserichards replied:
no longer applicable, this piece of code is gone now :)

nickbrandt commented:
@eliserichards saw I was requested for review again, though appears nothing was changed? Did I miss something

devinreams commented:
@mozilla-lockbox/ux We confirmed the last UX finding was addressed:

![img_1092](https://user-images.githubusercontent.com/49511/49972664-94affb00-feef-11e8-9915-d7988d381a44.jpeg)


... and asked @eliserichards to go ahead and merge knowing any other issues can and should be filed and fixed immediately once you're back and able to test again.

eliserichards commented:
@mozilla-lockbox/ux: re Devin's comment above, I got the ok from Sasha and Devin that this issue is satisfied. That being said, can you please take a quick look when you get back and if there are any outstanding issues I can file them and get them cleaned up ASAP :)
